2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/activity/search/SearchActivity.kt

```diff
@@ -2,52 +2,95 @@ package com.example.architecture.activity.search
 
 class SearchActivity : AppCompatActivity() {
 
-    private val vm = SearchViewModel(NaverRepositoryImpl(this))
+    @Suppress("UNCHECKED_CAST")
+    private val vm: SearchViewModel by viewModels {
+        object : ViewModelProvider.Factory {
+            override fun <T : ViewModel?> create(modelClass: Class<T>): T {
+
+                val naverLocalDataSource = NaverLocalDataSourceImpl.getInstance(applicationContext)
+                val naverRemoteDataSource = NaverRemoteDataSourceImpl
+                val naverRepository = NaverRepositoryImpl.getInstance(naverLocalDataSource, naverRemoteDataSource)
+
+                val resourceProvider = ResourceProviderImpl(this@SearchActivity.applicationContext)
+
+                return SearchViewModel(
+                    naverRepository,
+                    resourceProvider
+                ) as T
+            }
+        }
+    }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
-        val binding: ActivitySearchBinding =
-            DataBindingUtil.setContentView(this, R.layout.activity_search)
+        val binding: ActivitySearchBinding = DataBindingUtil.setContentView(this, R.layout.activity_search)
 
         binding.vm = vm
+        binding.lifecycleOwner = this
 
         setupRecyclerview()
-        setupViewModelEvent()
-
+        setupViewModelObserve()
     }
 
     private fun setupRecyclerview() {
         rv_search_movieList.adapter = MovieAdapter()
     }
 
-    private fun setupViewModelEvent() {
-        vm.showMessageEmptyResult.addOnPropertyChangedCallback(object :
-            OnPropertyChangedCallback() {
-            override fun onPropertyChanged(sender: Observable?, propertyId: Int) {
-                Toast.makeText(applicationContext, getString(R.string.not_found_result), Toast.LENGTH_SHORT).show()
-            }
-        })
+    private fun setupViewModelObserve() {
+        showToast()
+        searchMovieAfterMoment()
+    }
 
-        vm.showMessageEmptyKeyword.addOnPropertyChangedCallback(object :
-            OnPropertyChangedCallback() {
-            override fun onPropertyChanged(sender: Observable?, propertyId: Int) {
-                Toast.makeText(applicationContext, getString(R.string.empty_keyword), Toast.LENGTH_SHORT).show()
-            }
-        })
+    private fun showToast() {
+        vm.toastMessage
+            .observe(this, Observer { message ->
+                Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
+            })
     }
 
+    private fun searchMovieAfterMoment() {
+        vm.keyword
+            .debounce(AUTO_SEARCH_TIME)
+            .observe(this, Observer {
+                vm::searchMovie.invoke()
+            })
+    }
+
+    override fun onResume() {
+        super.onResume()
+        bindViewModel()
+    }
+
+    private fun bindViewModel() {
+        vm.bindViewModel()
+    }
+
+    override fun onPause() {
+        unbindViewModel()
+        super.onPause()
```

Comment:
```suggestion
    override fun onPause() {
    vm.unbindViewModel()
        super.onPause()
```
이 좋을 것 같습니다. (1) 방법
그런데 좀 이상한 것이 onPause에 선언하면 다른 엑티비티가 있을 경우에 다시 back하면 onCreate를 안타기 때문에 안전한 코드가 아닌 것 같습니다.

Reply:
onCreate에 bind 되는 것이 아닌 onResume()에 다시 bindviewModel하는 것으로 되어있어서 괜찮을 것 같습니다!

unbindViewModel로 바로 써도 되지만,
vm의 있는 로직을 바로 호출하는게 아니라 unbind하는 작업을한다는것을 표현하는게 더 좋을 것 같다는 생각입니다.

Reply:
의도하신거라면 onCreate가 아닌 반복적인 bind때문에 굳이 onResume에서 해야할 이유가 있나요? onResume이 back했을때도 작동되는걸 의도하신거라면 flag를 두어서 처음엔 onCreate 그 이후엔 onResume에서 bind하는 쪽으로 코드를 작성하는게 좋을 것 같습니다.

Reply:
>  flag를 두어서 처음엔 onCreate 그 이후엔 onResume에서 bind하는 쪽으로 코드를 작성하는게 좋을 것 같습니다.

 의도를 잘 모르겠습니다. onCreate이후 에도 onResume을 타는데 플래그를 사용해서 굳이 제어를 하려는 의도를 잘 모르겠습니다.